blogs/2023/07/20/mangling-vscode.md

```diff
@@ -0,0 +1,177 @@
+
+## False starts with private props
+
+Looking back over our minified sources, another thing that jumped out at me was how many long names I saw starting with `_`. By convention, this indicates a private property. Surely private properties can be safely mangled and code outside of the class would be none the wiser, right? And wait, shouldn't esbuild be doing this already for us? Yet I knew that the folks who wrote esbuild are no slouches. If esbuild wasn't mangling private properties, it was almost certainly for good reason.
+
+As I thought about the problem more, I realized that private properties are affected by the same dynamic property lookup issue shown in the `longPropertyName` example above. I'm sure a smart TypeScript programmer like yourself would never write such code, but dynamic patterns are common enough in real world codebases that esbuild opts to play it safe.
+
+Also keep in mind that the `private` keyword in TypeScript is really just a polite suggestion. When TypeScript code is compiled to JavaScript, the `private` keyword is basically removed. That means there's nothing to stop rude code outside of the class from reaching in and accessing private properties willy-nilly:
+
+```ts
+class Foo {
+    private bar = 123;
+}
+
+const foo: any = new Foo();
+console.log(foo.bar);
+```
+
+Hopefully your code isn't doing crazy stuff like this directly, but carelessly changing property names can potentially bite you in plenty of fun unexpected ways such as with object spreads, serialization, and when distinct classes share common property names.
+
+Thankfully I realized that with VS Code I had one huge advantage: I was working with a (mostly) sane codebase. I could make many assumptions that esbuild couldn't, such as that there are no dynamic private properties accesses or bad `any` accesses. This further simplified the problem I was facing.
+
+So together Johannes Rieken ([@johannesrieken](https://twitter.com/johannesrieken)) and I started to explore private property mangling. Our first idea was to try adopting JavaScript's native `#private` fields everywhere in our codebase. Not only are private fields immune to all of the problems detailed above, they already get mangled automatically by esbuild. Moving closer to plain old JavaScript was also appealing.
+
+However, we quickly dismissed this approach as it would require massive (meaning risky) code changes, including removing all of our uses of [parameter properties](https://www.typescriptlang.org/docs/handbook/2/classes.html#parameter-properties). As a relatively new feature, private fields also haven't been optimized across all runtimes yet. Using them can introduce slowdowns ranging from negligible to around 30%. Although this may be the correct change in the long run, it wasn't what we needed right now.
```

Comment:
For Safari it's actually worst than 30%: https://bugs.webkit.org/show_bug.cgi?id=258660